internal/tokens/https/https.go

```diff
@@ -0,0 +1,223 @@
+	"github.com/hyperledger-labs/firefly/internal/config"
+	"github.com/hyperledger-labs/firefly/internal/i18n"
+	"github.com/hyperledger-labs/firefly/internal/log"
+	"github.com/hyperledger-labs/firefly/internal/restclient"
+	"github.com/hyperledger-labs/firefly/internal/wsclient"
+	"github.com/hyperledger-labs/firefly/pkg/fftypes"
+	"github.com/hyperledger-labs/firefly/pkg/tokens"
+)
+
+type HTTPS struct {
+	ctx          context.Context
+	capabilities *tokens.Capabilities
+	callbacks    tokens.Callbacks
+	client       *resty.Client
+	wsconn       wsclient.WSClient
+	closed       chan struct{}
+}
+
+type wsEvent struct {
+	Event msgType            `json:"event"`
+	ID    string             `json:"id"`
+	Data  fftypes.JSONObject `json:"data"`
+}
+
+type msgType string
+
+const (
+	messageReceipt   msgType = "receipt"
+	messageTokenPool msgType = "token-pool"
+)
+
+type responseData struct {
+	RequestID string `json:"id"`
+}
+
+type createPool struct {
+	ID        *fftypes.UUID     `json:"client_id"`
+	Type      fftypes.TokenType `json:"type"`
+	Namespace string            `json:"namespace"`
+	Name      string            `json:"name"`
+}
+
+func (h *HTTPS) Name() string {
+	return "https"
+}
+
+func (h *HTTPS) Init(ctx context.Context, prefix config.Prefix, callbacks tokens.Callbacks) (err error) {
+
+	h.ctx = log.WithLogField(ctx, "proto", "https")
+	h.callbacks = callbacks
+
+	if prefix.GetString(restclient.HTTPConfigURL) == "" {
+		return i18n.NewError(ctx, i18n.MsgMissingPluginConfig, "url", "tokens.https")
+	}
+
+	h.client = restclient.New(h.ctx, prefix)
+	h.capabilities = &tokens.Capabilities{}
+
+	if prefix.GetString(wsclient.WSConfigKeyPath) == "" {
+		prefix.Set(wsclient.WSConfigKeyPath, "/api/ws")
+	}
+	h.wsconn, err = wsclient.New(ctx, prefix, nil)
+	if err != nil {
+		return err
+	}
+
+	h.closed = make(chan struct{})
+	go h.eventLoop()
+
+	return nil
+}
+
+func (h *HTTPS) Start() error {
+	return h.wsconn.Connect()
+}
+
+func (h *HTTPS) Capabilities() *tokens.Capabilities {
+	return h.capabilities
+}
+
+func (h *HTTPS) handleReceipt(ctx context.Context, data fftypes.JSONObject) error {
+	l := log.L(ctx)
+
+	requestID := data.GetString("id")
+	success := data.GetBool("success")
+	message := data.GetString("message")
+	if requestID == "" {
+		l.Errorf("Reply cannot be processed: %+v", data)
+		return nil // Swallow this and move on
+	}
+	updateType := fftypes.OpStatusSucceeded
+	if !success {
+		updateType = fftypes.OpStatusFailed
+	}
+	l.Infof("Tokens '%s' reply (request=%s) %s", updateType, requestID, message)
+	return h.callbacks.TokensTxUpdate(h, requestID, updateType, message, data)
+}
+
+func (h *HTTPS) handleTokenPoolCreate(ctx context.Context, data fftypes.JSONObject) (err error) {
+	ns := data.GetString("namespace")
+	name := data.GetString("name")
+	clientID := data.GetString("client_id")
+	tokenType := data.GetString("type")
+	poolID := data.GetString("pool_id")
+	authorAddress := data.GetString("author")
```

Comment:
As much as a note to self of a hard problem to think about as we work through this, but I'm wondering how the difference between blockchain Identity and FireFly Identity (think a DID here once we've got a useful Identity plugin built) is managed.

e.g. if someone else in the network creates a token pool, what identities are stored where, and are we consistent in how we name those identities on the API properties.

Feel it's super easy for us to miss this, in the case where there's no translation (so it happens to be the case today that the blockchain identity and the FireFly identity are the same string).

Reply:
Agreed; it's hard to mentally think about the separation when they're currently equivalent.